egs/librispeech/asr/simple_v1/mmi_mbr_train.py

```diff
@@ -0,0 +1,515 @@
+    feature = batch['features']
+    supervisions = batch['supervisions']
+    supervision_segments = torch.stack(
+        (supervisions['sequence_idx'],
+         torch.floor_divide(supervisions['start_frame'],
+                            model.subsampling_factor),
+         torch.floor_divide(supervisions['num_frames'],
+                            model.subsampling_factor)), 1).to(torch.int32)
+    indices = torch.argsort(supervision_segments[:, 2], descending=True)
+    supervision_segments = supervision_segments[indices]
+
+    texts = supervisions['text']
+    texts = [texts[idx] for idx in indices]
+    assert feature.ndim == 3
+    # print(supervision_segments[:, 1] + supervision_segments[:, 2])
+
+    feature = feature.to(device)
+    # at entry, feature is [N, T, C]
+    feature = feature.permute(0, 2, 1)  # now feature is [N, C, T]
+    if is_training:
+        nnet_output = model(feature)
+    else:
+        with torch.no_grad():
+            nnet_output = model(feature)
+
+    # nnet_output is [N, C, T]
+    nnet_output = nnet_output.permute(0, 2, 1)  # now nnet_output is [N, T, C]
+
+    if is_training:
+        num, den, mbr_num, mbr_den = graph_compiler.compile(texts, P)
+    else:
+        with torch.no_grad():
+            num, den, mbr_num, mbr_den = graph_compiler.compile(texts, P)
+
+    assert num.requires_grad == is_training
+    assert den.requires_grad is False
+    assert mbr_num.requires_grad is False
+    assert mbr_den.requires_grad is False
+
+    num = num.to(device)
+    den = den.to(device)
+
+    mbr_num = mbr_num.to(device)
+    mbr_den = mbr_den.to(device)
+
+    dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+    assert nnet_output.device == device
+
+    logging.info('intersect dense num')
+    num_graph = k2.intersect_dense(num, dense_fsa_vec, 10.0)
+
+    logging.info('intersect dense den')
+    den_graph = k2.intersect_dense(den, dense_fsa_vec, 10.0)
+
+    logging.info('num get tot scores')
+    num_tot_scores = num_graph.get_tot_scores(log_semiring=True,
+                                              use_double_scores=True)
+
+    logging.info('den get tot scores')
+    den_tot_scores = den_graph.get_tot_scores(log_semiring=True,
+                                              use_double_scores=True)
+
+    tot_scores = num_tot_scores - den_scale * den_tot_scores
+
+    (tot_score, tot_frames,
+     all_frames) = get_tot_objf_and_num_frames(tot_scores,
+                                               supervision_segments[:, 2])
+
+    logging.info('mbr_num_graph')
+    # now for mbr loss
+    mbr_num_graph = k2.intersect_dense(mbr_num,
+                                       dense_fsa_vec,
+                                       10.0,
+                                       seqframe_idx_name='seqframe_idx')
+    logging.info('mbr_den_graph {}'.format(mbr_den.shape))
+
+    mbr_den_graph = k2.intersect_dense_pruned(mbr_den,
```

Comment:
The training gets stuck here. `k2.intersect_dense_prune` seems never terminating.
I have decreased search beam and max active states.

Reply:
mm I may need to debug this.